### ISSUE TITLE
BPMN20 XML import throws NullPointerException when no process is found

### ISSUE THREAD

**Issue by grexe** [flowable/flowable-engine#2065]
**Describe the bug**
I have designed a BPMN2.0 process model in Signavio modeller and exported it as BPMN 2.0 XML. When trying to import the XML into Flowable 6.4.2, the importer throws a NPE:
```
2019-12-17 15:36:03.256 ERROR 1 --- [nio-8080-exec-8] o.f.u.m.s.FlowableModelQueryService      : Import failed for BigPicture.bpmn

org.flowable.bpmn.exceptions.XMLException: Error processing BPMN document
	at org.flowable.bpmn.converter.BpmnXMLConverter.convertToBpmnModel(BpmnXMLConverter.java:448) ~[flowable-bpmn-converter-6.4.2.1.jar:6.4.2.1]
	at org.flowable.ui.modeler.service.FlowableModelQueryService.importProcessModel(FlowableModelQueryService.java:179) ~[flowable-ui-modeler-logic-6.4.2.1.jar:6.4.2.1]
...
Caused by: java.lang.NullPointerException
	at org.flowable.bpmn.converter.parser.ExtensionElementsParser.parse(ExtensionElementsParser.java:55) ~[flowable-bpmn-converter-6.4.2.1.jar:6.4.2.1]
	at org.flowable.bpmn.converter.BpmnXMLConverter.convertToBpmnModel(BpmnXMLConverter.java:399) ~[flowable-bpmn-converter-6.4.2.1.jar:6.4.2.1]
```

**Expected behavior**
The model should be imported and visualised. It works in e.g. https://bpmn.io online demo.

**Code**
I will try to provide a simple test case for reproducing this bug if needed.

**Additional context**
Using flowable 6.4.2.1 (also tried 6.5.0-SNAPSHOT as of today).
Checked out source and ran the `org.flowable.engine.test.bpmn.usertask.ImportExportTest` with my file.
The problem is that if no process is found in the XML (there is none defined), then in `org.flowable.bpmn.converter.parser.ExtensionElementsParser:parse()`, the if/else block does not set the `parentElement` (`parentElement = activeProcess` will set `null`), and the subsequent call `parentElement.addExtensionElement(extensionElement);`will fail, resulting in a NPE.

**Comment by tstephen** [comment]
Could you attach the `.bpmn` file (if possible)?

**Comment by tstephen** [comment]
And if that's a problem... If there is no process in the BPMN, what _is_ in there?

**Comment by grexe** [comment]
ok I've added a simple check and create a new, empty `Process` as parent now if none is found.
This works, and I get the 2 processes that are actually in the model: "system" and "customer".

However, now the deployment fails with loads of parse errors (after everything has actually been successfully converted from model to XML and backwards in the test?!):
```
org.flowable.common.engine.api.FlowableException: Errors while parsing:
[Validation set: 'flowable-executable-process' | Problem: 'flowable-sendtask-invalid-implementation'] : One of the attributes 'type' or 'operation' is mandatory on sendTask - [Extra info : processDefinitionId = sid-B72004ED-2165-44C6-BA4A-0D850E6A865D | processDefinitionName = System |  | id = sid-54A94553-EF8A-4C54-B43D-237AE1D67F3B |  | activityName = WorkItem fÃÂ¼r Sachbearbeiter anlegen | ] ( line: 478, column: 107)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-sendtask-invalid-implementation'] : One of the attributes 'type' or 'operation' is mandatory on sendTask - [Extra info : processDefinitionId = sid-B72004ED-2165-44C6-BA4A-0D850E6A865D | processDefinitionName = System |  | id = sid-59821A89-4E4B-4A8E-96E5-BA4450218D64 |  | activityName = SachverstÃÂ¤ndingen beauftragen | ] ( line: 530, column: 101)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-sendtask-invalid-implementation'] : One of the attributes 'type' or 'operation' is mandatory on sendTask - [Extra info : processDefinitionId = sid-B72004ED-2165-44C6-BA4A-0D850E6A865D | processDefinitionName = System |  | id = sid-9F1AB844-3BA1-4859-BB12-416A01B75FFC |  | activityName = Information anfordern | ] ( line: 589, column: 90)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-sendtask-invalid-implementation'] : One of the attributes 'type' or 'operation' is mandatory on sendTask - [Extra info : processDefinitionId = sid-B72004ED-2165-44C6-BA4A-0D850E6A865D | processDefinitionName = System |  | id = sid-35F8A584-013B-4D8A-BCA2-D6DEF38246D7 |  | activityName = Informationen anfordern | ] ( line: 745, column: 92)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-message-event-missing-message-ref'] : attribute 'messageRef' is required - [Extra info : processDefinitionId = sid-E2EF1A3C-B582-4B78-ACBC-6A9EACBEA9BB | processDefinitionName = Kunde |  | id = sid-26F358CC-8F01-45FB-AA81-4FF9C5656BE0 |  | activityName = Kunde reicht Schaden ein | ] ( line: 85, column: 107)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-event-timer-missing-configuration'] : Timer needs configuration (either timeDate, timeCycle or timeDuration is needed) - [Extra info : processDefinitionId = sid-B72004ED-2165-44C6-BA4A-0D850E6A865D | processDefinitionName = System |  | id = sid-DB15AFE1-EE63-4EFE-BCB9-0839222103BF |  | activityName = Nach Fristablauf Fall abschlieÃÂen  (3 Jahre) | ] ( line: 619, column: 130)
[Validation set: 'flowable-executable-process' | Problem: 'flowable-process-definition-not-executable'] : All process definition are set to be non-executable (property 'isExecutable' on process). This is not allowed. - [Extra info : ]
[Validation set: 'flowable-executable-process' | Problem: 'flowable-throw-event-invalid-eventdefinition'] : Unsupported intermediate throw event type - [Extra info : processDefinitionId = sid-E2EF1A3C-B582-4B78-ACBC-6A9EACBEA9BB | processDefinitionName = Kunde |  | id = sid-26F358CC-8F01-45FB-AA81-4FF9C5656BE0 |  | activityName = Kunde reicht Schaden ein | ] ( line: 85, column: 107)

	at org.flowable.engine.impl.bpmn.parser.BpmnParse.execute(BpmnParse.java:177)
	at org.flowable.engine.impl.bpmn.deployer.ParsedDeploymentBuilder.createBpmnParseFromResource(ParsedDeploymentBuilder.java:97)
	at org.flowable.engine.impl.bpmn.deployer.ParsedDeploymentBuilder.build(ParsedDeploymentBuilder.java:55)
	at org.flowable.engine.impl.bpmn.deployer.BpmnDeployer.deploy(BpmnDeployer.java:78)
	at org.flowable.engine.impl.persistence.deploy.DeploymentManager.deploy(DeploymentManager.java:62)
	at org.flowable.engine.impl.cmd.DeployCmd.executeDeploy(DeployCmd.java:126)
	at org.flowable.engine.impl.cmd.DeployCmd.execute(DeployCmd.java:69)
```

**Comment by grexe** [comment]
> Could you attach the `.bpmn` file (if possible)?

I have to find a minimal non-working definition as this is a real company internal process. I hope we can pinpoint the problem without that.
It seems the checks are much stricter in flowable than in e.g. bpmn.io, is there a "lenient" option?,-)

**Comment by grexe** [comment]
PS: my change that got the model through (but probably ruined it in other ways) was:
```
public class ExtensionElementsParser implements BpmnXMLConstants {

    public void parse(XMLStreamReader xtr, List<SubProcess> activeSubProcessList, Process activeProcess, BpmnModel model) throws Exception {
        BaseElement parentElement;
        if (!activeSubProcessList.isEmpty()) {
            parentElement = activeSubProcessList.get(activeSubProcessList.size() - 1);

        } else {
            if (activeProcess != null) {
                parentElement = activeProcess;
            } else {
                parentElement = new Process();
            }
        }
```

**Comment by tstephen** [comment]
There's 3 tools in play here: signavio, bpmn.io and flowable. In my experience it is not so much that one is stricter than the other but they are just different. In particular, flowable is orientated more towards execution than the other two.

From your description and the code change I would guess you perhaps have a collaboration with one process that is either incomplete or maybe not even intended for execution. Minimally, something like this:

![Selection_034](https://user-images.githubusercontent.com/1498508/71020964-3bc40580-20f5-11ea-9d5b-fd219588ccb2.png)

There's nothing _wrong_ with this kind of process model but it won't execute and because Flowable is focused on execution you may encounter this kind of issue. If you do put together a minimal example I'll probably be able to say more about the missing details.

FYI the pic above came from this model:
```
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="Definitions_0uur48e" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="3.4.1">
  <bpmn:collaboration id="Collaboration_00ykapx">
    <bpmn:participant id="Participant_1mp7de9" name="System" processRef="Process_149coqt" />
    <bpmn:participant id="Participant_1hwx5oq" name="Customer" processRef="Process_163abe9" />
  </bpmn:collaboration>
  <bpmn:process id="Process_149coqt" isExecutable="true">
    <bpmn:startEvent id="StartEvent_1">
      <bpmn:outgoing>SequenceFlow_1rx05hi</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:endEvent id="EndEvent_061z8p5">
      <bpmn:incoming>SequenceFlow_10t5x5v</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:userTask id="Task_0hydd5c" name="Do stuff">
      <bpmn:incoming>SequenceFlow_1rx05hi</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_10t5x5v</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:sequenceFlow id="SequenceFlow_1rx05hi" sourceRef="StartEvent_1" targetRef="Task_0hydd5c" />
    <bpmn:sequenceFlow id="SequenceFlow_10t5x5v" sourceRef="Task_0hydd5c" targetRef="EndEvent_061z8p5" />
  </bpmn:process>
  <bpmn:process id="Process_163abe9" isExecutable="false" />
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Collaboration_00ykapx">
      <bpmndi:BPMNShape id="Participant_1mp7de9_di" bpmnElement="Participant_1mp7de9" isHorizontal="true">
        <dc:Bounds x="129" y="79" width="600" height="250" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
        <dc:Bounds x="192" y="192" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Participant_1hwx5oq_di" bpmnElement="Participant_1hwx5oq" isHorizontal="true">
        <dc:Bounds x="129" y="380" width="600" height="250" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_1rx05hi_di" bpmnElement="SequenceFlow_1rx05hi">
        <di:waypoint x="228" y="210" />
        <di:waypoint x="280" y="210" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="EndEvent_061z8p5_di" bpmnElement="EndEvent_061z8p5">
        <dc:Bounds x="432" y="192" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_10t5x5v_di" bpmnElement="SequenceFlow_10t5x5v">
        <di:waypoint x="380" y="210" />
        <di:waypoint x="432" y="210" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="UserTask_1sjm2zs_di" bpmnElement="Task_0hydd5c">
        <dc:Bounds x="280" y="170" width="100" height="80" />
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```

**Comment by grexe** [comment]
OK I was able to reproduce the error and also my quick fix with a completely stripped down version of our business process, also this error confirms the point about the process not being executable that @tstephen made:
```
[Validation set: 'flowable-executable-process' | Problem: 'flowable-process-definition-not-executable'] : All process definition are set to be non-executable (property 'isExecutable' on process). This is not allowed. - [Extra info : ]
```

**Comment by grexe** [comment]
So I got my test to work, I am not sure but I think Flowable parses the model not quite right:
1. 'BpmnXMLConverter' expects the `process` element to occur first, before any `extensionElements`: this was the root cause for this issue, causing the `activeProcess` not being set before parsing ´extensionElements` block, resulting in a `NullPointerException` there.
This assumption is not backed by the BPMN spec afaik and certainly not respected by commonly used tools like bpmn.io modeller.
1. after moving the process definition up in my example, I noticed that the process block is parsed *3* times, didn't yet discover why, as it appears only once in the XML.
1. I got some model errors on deployment that were not raised on parsing the model from file, e.g. I had to add `messageRef` attributes to my `messageEventDefinition` elements, although not required by the XSD.
1. my `intermediateThrowEvent`'s of type `MessageEventDefinition` (implicitly set, not in my XML, but I guess it's deferred from the child element `messageEventDefinition`?) was not detected as a valid type by the `IntermediateThrowEventValidator`, so I added that to the `instanced` checks

After all this, my test deployment worked:)

I will now prepare a PR for review and attach my test file if you agree with my findings.

**Comment by grexe** [comment]
[bpmn-samples.zip](https://github.com/flowable/flowable-engine/files/3979642/bpmn-samples.zip)
2 testfiles:
1. `diagram.bpmn20.xml`: as originally exported by bpmn.io
1. `diagram-executable.bpmn20.xml`: adapted for flowable (also changed some needed IDs for process, messages and events to be more human readable for testing)

**Comment by tijsrademakers** [comment]
Like Tim already explained very well the Flowable Modeler tool focuses on executable process models. So having a process model with multiple pools and one pool not being executable is not supported. The goal of our Modeler is to, in the end, run the process model on the Flowable BPMN engine.

A similar thing is the case for the message event definition without a message ref. For an executable process model the message ref is required, because that defines the type of message you are sending.

For the intermediate message throw event I already explained on the PR that you created that we don't support this because we don't support collaboration diagrams with message interaction. Instead, signal events can be used to communicate between related process models to provide a way to communicate events from one process model to the other.

**Comment by tstephen** [comment]
Further to Tijs' points about the differences between an execution engine and modeling... 

Nonetheless, thanks for working out the minimal test case Gregor. My interest was piqued as it generally is for issues of interoperability between different vendors. 

To your points:
1. It's not the order of process and collaboration elements itself that proves to be the problem. It is because you have extension elements within the collaboration. Since Flowable has no use for these (they are Signavio extensions) the parser has not anticipated it. Your change in the second file means that because the process has already been read the extensions are then readable.
2. I did not find the repeated parsing in my test.
3. XSD validation is `necessary but not sufficient`. If you use a message event then by execution time it must be fully specified. Obviously for modeling tools this is not necessary and indeed it may be desirable to exchange incomplete models or models not destined for execution. The spec talks about this in terms of compliance levels. In essence, a compliant modeling tool has no obligation to care about execution elements.
4. Yes the child element `messageEventDefinition` is what makes the event a message one.

PR for #1 on its way. Given the execution conversation I simply discarded the extension elements but I may see if a slightly more intelligent way is possible if I have some further time later on.

**Comment by tstephen** [comment]
Sorry for not spotting https://github.com/flowable/flowable-engine/pull/2068 before I did this.

**Comment by grexe** [comment]
Thanks @tijsrademakers and @tstephen for the thorough and patient explanations of all the factors at play here; quite a lot to digest since only now I finally have a real use case in a real project, previously only loosely followed the workflow topic out of interest.

I'd be very happy if the PR #2068 would be included in the near future, as I guess this could happen quite often when designing workflows with business experts and later moving to implementation without friction and the need for manual changes to the BPMN XML.
